omnigibson/action_primitives/action_primitive_set_base.py

```diff
@@ -0,0 +1,76 @@
+import inspect
+from abc import ABCMeta, abstractmethod
+from enum import IntEnum
+from typing import List
+
+from future.utils import with_metaclass
+
+from omnigibson.robots import BaseRobot
+from omnigibson.scenes.interactive_traversable_scene import InteractiveTraversableScene
+from omnigibson.tasks.task_base import BaseTask
+
+REGISTERED_PRIMITIVE_SETS = {}
+
+class ActionPrimitiveError(ValueError):
+    class Reason(IntEnum):
+        # The planning for a primitive was successfully completed, but an error occurred during execution.
+        EXECUTION_ERROR = 0
+
+        # The planning for a primitive failed possibly due to not being able to find a path.
+        PLANNING_ERROR = 1
+
+        # A primitive could not be executed because a precondition was not satisfied, e.g. PLACE was called without an
+        # object currently in hand.
+        PRE_CONDITION_ERROR = 2
+
+        # A sampling error occurred: a position to place an object could not be found.
+        SAMPLING_ERROR = 3
+
+        # The execution of the primitive happened correctly, but while checking post-conditions, an error was found.
+        POST_CONDITION_ERROR = 4
+
+    def __init__(self, reason: Reason, message, metadata=None):
+        self.reason = reason
+        self.metadata = metadata if metadata is not None else {}
+        super().__init__(f"{reason.name}: {message}. Additional info: {metadata}")
+
+
+class ActionPrimitiveErrorGroup(ValueError):
+    def __init__(self, exceptions: List[ActionPrimitiveError]) -> None:
+        self._exceptions = tuple(exceptions)
+        submessages = [f"Attempt {i}: {e}" for i, e in enumerate(exceptions)]
+        submessages = "\n\n".join(submessages)
+        message = "An error occurred during each attempt of this action.\n\n" + submessages
+        super().__init__(message)
+
+    @property
+    def exceptions(self):
+        return self._exceptions
+
+
+class BaseActionPrimitiveSet(with_metaclass(ABCMeta, object)):
+    def __init_subclass__(cls, **kwargs):
+        """
+        Registers all subclasses as part of this registry. This is useful to decouple internal codebase from external
+        user additions. This way, users can add their custom robot by simply extending this Robot class,
+        and it will automatically be registered internally. This allows users to then specify their robot
+        directly in string-from in e.g., their config files, without having to manually set the str-to-class mapping
+        in our code.
+        """
+        if not inspect.isabstract(cls):
+            REGISTERED_PRIMITIVE_SETS[cls.__name__] = cls
+
+    def __init__(self, task, scene, robot):
```

Comment:
this looks suspiciously like an `Environment` class instance. Is there a reason why that class couldn't be extended / wrapped directly?

Reply:
Pass the env as an argument rather than passing in task, scene, robot.